STM32WB: Add HCI driver and enable BLE

erwango opened the pull request:
This series enable BLE on STM32WB.

It includes flash driver for persistent storage.

This is based on https://github.com/zephyrproject-rtos/zephyr/pull/16373

erwango committed:
ext/hal/st/lib: stm32wb hci shared ram driver

Library providing hci driver for host/controller communication
over shared RAM

Origin: ST Microelectronics
License: BSD-3-Clause
URL: http://www.st.com/en/embedded-software/stm32cubewb.html
Commit: 1.0.0
Purpose: Shared RAM driver for STM32WB
Maintained-by: External

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/bluetooth/hci: Implement HCI driver for stm32wb

Implement HCI driver for STM32WB. It allows host to controller.
It is based on ST library allowing communication over RAM shared
bewteen chip's C-M4 and C-M0 cores.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_wb55rg: Configure HCI driver

Add requested configuration to enable HCI driver

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_wb55rg: Configure flash partitions

Configure flash partitions on nucleo_wb55rg

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/flash: stm32wb: Add HSEM

Add HSEM for Flash access since shared between 2 cores.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: stm32wb: Limit flash size to 812K

stm32wb55rg is a dual core SoC (M4/M0) with a radio/protocol stack
running on M0 core. M0 FW starts at flash address 0x80CB000.
So flash size available to zephyr is 812K instead of 1M.
Configure package size to reflect this (flash size is now 812K)
and update MPU setting to exclude M0 FW. From MPU point of view,
we set flash size to 512K since we should chose a size being a power
of 2.
If we consider there will be 2 images partitions in a FOTA scheme,
512 K is beyond image-0 and hence is sufficient to cover application
image.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

jfischer-phytec-iot commented:
@erwango Have you tried it with `samples/bluetooth/peripheral*` ? I noticed that RF Performance is not that good , 30dBm ... 40dBm worse than nRF52 DK boards or FRDM-KW41Z.

erwango commented:
> @erwango Have you tried it with samples/bluetooth/peripheral* ? I noticed that RF Performance is not that good , 30dBm ... 40dBm worse than nRF52 DK boards or FRDM-KW41Z.

Yes I have tried it but didn't pay attention, I can see now that RF performance is not quite good indeed. Though, with other fw this is working correctly, so I may have messed up with SET_TX_POWER command. Let me have a check.

erwango commented:
@jfischer-phytec-iot, it should be fixed now.

jfischer-phytec-iot commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -0,0 +1,398 @@
+
+	switch (hcievt->evtserial.type) {
+	case HCI_EVT:
+		BT_DBG("EVT: hcievt->evtserial.evt.evtcode: %d",
```
suggestion, use `BT_DBG("EVT: hcievt->evtserial.evt.evtcode: 0x%02x"` to have it equal to `hci_core `

erwango replied:
Txs, will take it into account at next push

jfischer-phytec-iot commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -0,0 +1,398 @@
+		switch (hcievt->evtserial.evt.evtcode) {
+		case BT_HCI_EVT_VENDOR:
+			/* Vendor events are currently unsupported */
+			BT_ERR("Unknown evtcode type %d",
```
suggestion, use `BT_ERR("Unknown evtcode type 0x%02x"` to have it equal to hci_core

jfischer-phytec-iot commented:
@erwango I tested few peripheral samples (peripheral and peripheral_sc_only). The connection was not successful, log part from peripheral sample:

```
[00:00:28.727,176] <dbg> bt_hci_core.hci_rx_thread: calling fifo_get_wait
[00:00:29.042,027] <dbg> hci_ipm.TM_EvtReceivedCb: EVT: hcievt->evtserial.evt.evtcode: 0x01
[00:00:29.042,061] <dbg> bt_hci_core.bt_recv: buf 0x20002234 len 42
[00:00:29.042,114] <dbg> bt_hci_core.hci_rx_thread: buf 0x20002234 type 1 len 42
[00:00:29.042,122] <dbg> bt_hci_core.hci_event: event 0x01
[00:00:29.042,171] <wrn> bt_hci_core: Unhandled event 0x01 len 40: 0b0007000400100100ffff002800000~
[00:00:29.042,200] <dbg> bt_hci_core.hci_rx_thread: calling fifo_get_wait
[00:00:29.222,009] <dbg> hci_ipm.TM_EvtReceivedCb: EVT: hcievt->evtserial.evt.evtcode: 0x3e
[00:00:29.222,041] <dbg> bt_hci_core.bt_recv: buf 0x20002234 len 12
[00:00:29.222,093] <dbg> bt_hci_core.hci_rx_thread: buf 0x20002234 type 1 len 12
[00:00:29.222,102] <dbg> bt_hci_core.hci_event: event 0x3e
[00:00:29.222,125] <dbg> bt_hci_core.hci_le_meta_event: subevent 0x03
[00:00:29.222,135] <dbg> bt_hci_core.le_conn_update_complete: status 0, handle 2049
[00:00:29.222,145] <dbg> bt_conn.bt_conn_ref: handle 2049 ref 2
[00:00:29.222,156] <dbg> bt_conn.bt_conn_unref: handle 2049 ref 1
[00:00:29.222,185] <dbg> bt_hci_core.hci_rx_thread: calling fifo_get_wait
[00:00:33.961,587] <dbg> hci_ipm.TM_EvtReceivedCb: EVT: hcievt->evtserial.evt.evtcode: 0x01
[00:00:33.961,622] <dbg> bt_hci_core.bt_recv: buf 0x20002234 len 42
[00:00:33.961,675] <dbg> bt_hci_core.hci_rx_thread: buf 0x20002234 type 1 len 42
[00:00:33.961,683] <dbg> bt_hci_core.hci_event: event 0x01
[00:00:33.961,732] <wrn> bt_hci_core: Unhandled event 0x01 len 40: 0b0007000400100100ffff002800000~
[00:00:33.961,761] <dbg> bt_hci_core.hci_rx_thread: calling fifo_get_wait
[00:00:38.956,506] <dbg> hci_ipm.TM_EvtReceivedCb: EVT: hcievt->evtserial.evt.evtcode: 0x01
[00:00:38.956,540] <dbg> bt_hci_core.bt_recv: buf 0x20002234 len 42
[00:00:38.956,593] <dbg> bt_hci_core.hci_rx_thread: buf 0x20002234 type 1 len 42
[00:00:38.956,602] <dbg> bt_hci_core.hci_event: event 0x01
[00:00:38.956,651] <wrn> bt_hci_core: Unhandled event 0x01 len 40: 0b0007000400100100ffff002800000~
[00:00:38.956,680] <dbg> bt_hci_core.hci_rx_thread: calling fifo_get_wait
[00:00:42.526,520] <dbg> hci_ipm.TM_EvtReceivedCb: EVT: hcievt->evtserial.evt.evtcode: 0x3e
[00:00:42.526,553] <dbg> bt_hci_core.bt_recv: buf 0x20002234 len 12
[00:00:42.526,605] <dbg> bt_hci_core.hci_rx_thread: buf 0x20002234 type 1 len 12
[00:00:42.526,613] <dbg> bt_hci_core.hci_event: event 0x3e
[00:00:42.526,623] <dbg> bt_hci_core.hci_le_meta_event: subevent 0x03
[00:00:42.526,634] <dbg> bt_hci_core.le_conn_update_complete: status 0, handle 2049
[00:00:42.526,643] <dbg> bt_conn.bt_conn_ref: hDisconnected (reason 19)

```

erwango commented:
@jfischer-phytec-iot, I have no issue on peripheral on my side. From what I understand, you've already managed to get it working once, is that a new status? Or is this in particular conditions?
Regarding peripheral_sc_only, I don't know if this requires specific support from controller stack, I'll check this and get back to you. Thanks for interest!

jfischer-phytec-iot commented:
>From what I understand, you've already managed to get it working once, is that a new status? Or is this in particular conditions?

@erwango No, none of the peripheral samples work for me. Tested with nRF Connect and bluetoothctl on linux.

erwango commented:
> @erwango No, none of the peripheral samples work for me. Tested with nRF Connect and bluetoothctl on linux.

@jfischer-phytec-iot , ok. Well this is a big concern. Difference might come from the fact I use a pre mass market version. I'll get a board similar to yours and have a check.

erwango commented:
@jfischer-phytec-iot , I had a check and got things working on the nucleo version you should own (ES MB1355C). And peripheral is working on both ST BLE Profile and nRF Connect.

jfischer-phytec-iot commented:
@erwango How do you flash your device, I tried it over MSD support only.

erwango commented:
@jfischer-phytec-iot, yes I'm using mass storage as well. open-ocd support is upcoming but not yet available (AFAIK)

erwango commented:
@jfischer-phytec-iot , I'm sorry to say that, but would you consider having a try with stm32cube package?

jfischer-phytec-iot commented:
> I'm sorry to say that, but would you consider having a try with stm32cube package?

@erwango I have the package, but there is no way to build from command line?

erwango commented:
> @erwango I have the package, but there is no way to build from command line?

Using CubeMX Tool, you can generate a makefile. Though, it might not be direct.
Another option is I can generate send you a bin next week.

jfischer-phytec-iot commented:
>Another option is I can generate send you a bin next week.
:+1: Thanks

erwango committed:
drivers/flash: stm32wb: Add driver for stm32wb

Add flash driver for stm32wb

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
ext/hal: stm32wb: Configure M0 FW in LL only

In order to enable HCI mode in M0 firmware,
enable CFG_BLE_LL_ONLY in hci lib.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

dbkinder reviewed (approved):
+1 for doc changes

superna9999 commented on soc/arm/st_stm32/stm32wb/flash_registers.h:
```diff
@@ -0,0 +1,63 @@
+	volatile u32_t rsvd_3[6];
+	volatile u32_t sfr;
+	volatile u32_t srrvr;
+};
```
So you still add this structure, but you mostly use the HAL defines instead, except for common functions.

I'm dubious about this, but it's still correct...

erwango replied:
Yes, I wanted to limit the impacts on common functions.
How would you feel, in a next PR, to propagate the use of HAL defines and remove the need for defining these registers? Besides it seems that at least __flash_acr struct could be cleaned up since 'bit' part is unused if I'm correct (only 'val').

erwango replied:
@superna9999, any feedback ?

superna9999 commented on drivers/flash/flash_stm32.c:
```diff
@@ -34,19 +34,33 @@
 {
+
+#ifdef CONFIG_SOC_SERIES_STM32WBX
+	while (LL_HSEM_1StepLock(HSEM, CFG_HW_FLASH_SEMID))
```
I'm quite sure this could be defined as a standalone hardware semaphore driver and defined in DT instead of hardcoding usage here

erwango replied:
Agreed. Can we leave this as future work?
If you agree, I'll create an issue to track it.

erwango committed:
include/arch/arm: linker.ld: Add shared memory sections definitions

STM32WB HCI driver requires definition of 2 RAM regions to support
use of 3 shared memory sections: MAPPING_TABLE, MB_MEM1 and MB_MEM2.
In linker.ld, under conditions of HCI driver to be enabled,
define SRAM1 and SRAM2 based on input defined in stm32wb linker.
Then define the 3 sections MAPPING_TABLE, MB_MEM1 and MB_MEM

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: stm32wb: Delete useless sram definitions

sram2a and sram2b where defined in the prospect of defining
HCI shared memory sections from device tree.
Remove them, since this is not the case anymore.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
recheck

erwango commented:
@superna9999, removed calls to Cube HAL, can you review again?

erwango commented:
recheck

yhvh commented:
Hi, I'm also having issues with some Bluetooth samples with this patch, is there a set of tests to run or other debug that would help?

erwango commented:
@yhvh, on what samples are you facing issues?

yhvh commented:
@erwango I haven't been exhaustive, but I see issues with the peripheral* samples.
I am willing to do exhaustive tests, but I think we should be running integration tests so we are comparing apples to apples. I also offer to help write these.

Further - a testbed with a collection of supported boards would be nice.

erwango commented:
> @erwango I haven't been exhaustive, but I see issues with the peripheral* samples.

Can you be a little more specific on the issues you're seeing?
Which set-up are you using to test (Android app ?, R-PI, ..)? Are you able to connect (and get HR report for instance)?

> I am willing to do exhaustive tests, but I think we should be running integration tests so we are comparing apples to apples. I also offer to help write these.

Any help is welcome of course. Maybe the best approach is to submit a "RFC" issue describing your proposal.

> Further - a testbed with a collection of supported boards would be nice.

I guess this is a Zephyr generic topic. Also I would advice to create a dedicated issue ("Feature request" would fit).

yhvh commented:
bluetoothctl and nrf app on android, both fail to connect

erwango commented:
> bluetoothctl and nrf app on android, both fail to connect

Ok, I just tested it 2 hours back and was working. Also has been tested by others successfully.
Might come from your M0 f/w version that require an upgrade.
We know there are nucleo_wb samples in the wild that hold a M0 f/w not compatible with use on Zephyr. Can I ask you to create a dedicated issue for this and I'll try to support you on that track?

nashif commented:
tested with `samples/bluetooth/peripheral_hr`, works for me.

Needs rebase.

erwango commented:
@nashif , rebased

jfischer-no reviewed (approved):
LGTM, tested `peripheral_sc_only` and `peripheral_hr`

mtuxpe commented:
Hi there,
I am testing HCI driver using  WB55Nucleo and also P-NUCLEO-WB55 USB dongle.
Sometimes BLE stack doesn't start on dngle but I don't know how to reproduce yet.
I am developing a custom peripheral application to connect with ST BLE Sensor mobile.
To flash dongle on linux you can use STM32Programmer:
https://community.st.com/s/question/0D50X00009q3X4jSAE/stm32cubeprogrammer-for-linux
Use the bash script described on stefan1.5521405057577783E12 post.
Anyway good job @erwango

lst426 commented:
Why can't I configure Bluetooth to find out who's going to direct it? Thank you, God. Have you ever met?

mtuxpe commented:
Hi all,
Sometimes I got "Unhandled event" message while trying to connect to ST BLE Sensor Android app
On Zephyr side I am running  BLE peripheral.
How to get detailed log ?

[01:06:01.840,000] <inf> main: Connected
[01:06:02.287,000] <wrn> bt_hci_core: Unhandled event 0x02 len 40: 0b0007000400100d00ffff002800000~

erwango commented:
@mtuxpe, there's a dedicated application to trace HCI messages called monitor.
You can enable with CONFIG_BT_DEBUG_MONITOR=y, but this require to disable CONSOLE.
Then, you can get trace output over serial port with bluez btmon binary.

joerchan commented:
@mtuxpe It's possible that the HCI driver sent an event through bt_recv_prio, when it should have sent it through bt_recv.

pavlohamov commented:
@mtuxpe Hi. I've faces with the same issue. M0 firmware update - fixed the issue.

But now i'm seeing something strange with peripheral device. For all incoming connections the peer addresses are set to FF:FF:FF:FF:FF:FF, no matter how do i configure the BLE stack. for the Central role Addrs are ok. This issue block the secure connection.
@erwango Could you be so kid to point where to dig on this?

erwango commented:
@pavlohamov @mtuxpe, if you're facing a particular issue, please raise a new issue and provide details on your set up so we can look at it in details. Current PR is merged, we won't do much here.

PoojaKatariya commented:
Hello All,
 I downloaded the BLE_TransparentMode.hex file on the Nucleo board using STM32CubeProgrammer. Once the board was loaded with that hex file I tried to connect it with STM32CubeMonitor-RF. But while I'm trying to connect the device I'm getting a prompt which says "Device under test not responding COM port closed".

Also, before having received your support I tried to run the default application which comes in built with the board along with ST BLE Sensor mobile application. Then I tried to write the code as per the You Tube tutorial on STM32WB55 OLT session by STMicroelectronics and after which I flashed 
the board with stm32wb5x_BLE_Stack_fw.bin file as per one of the blogs on st.com. And now after having flashed the board; the blue tooth is not even getting detected on the sensor application.
Also, I am using STM32CubeIDE version 1.0.1 for all the development and not CubeMX platform.

Also I tried to boot up the Nucleo board right now as instructed in STM32WB package by:

a) Add a jumper between CN7.5 (VDD) and CN7.7 (BOOT0).
b) Power on the board via USB_USER and J2 jumper (USB_MCU)

Yet I am getting no results and facing the same issues as mentioned above.

erwango commented:
@PoojaKatariya, this github project is dedicated to zephyr RTOS, which is not the default application delivered with the board or in Cube package.
You will get proper support on st.community.com.
